(DOCS-4171) Rum event hierarchy (#15946)
DataDog/documentation: content/en/real_user_monitoring/guide/understanding-the-rum-event-hierarchy.md

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-switch-views.mp4" alt="RUM views" video="true" style="width:80%;">}}

Similarly to the `session.id`, each view will automatically have a unique `view.id` connected to it. 

### Actions, errors, resources, and long tasks

Within views, the SDK creates more granular events that all fall along the same level hierarchically. However, each event is unique and carries its own attributes and properties.

### Actions

Actions represent user activity on a page. In browsers, all click actions are automatically collected. On mobile, all taps, swipes, and scrolls are collected. Beyond these default actions, you can also send [custom actions][5], such as form completion and business transactions. 

#### Sample search: Top list of all "Add to cart" actions that led to an error 

> In this example, search within the actions event for an action error count (of at least 1) and specific action name to see how many times clicking "Add to cart" led to a user encountering an error.

Reviewer: As I re-read this line, I'm not sure it's super clear. How about this:
_In this example, search for all actions from users clicking on the "Add to cart" button that resulted in an error_

Author: Updated!

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-actions-all-add-to-cart.png" alt="Diagram of the RUM event hierarchy." style="width:80%;">}}

### Errors

> You can use RUM to collect [frontend errors][6] that occur during the user session. By default, the browser SDK creates error events for unhandled exceptions and console errors. Additionally, you can collect custom errors through the RUM addError API ([browser][7] | [mobile][8]). On mobile apps, you can also see if the error led to a session termination, also known as a crash.

Reviewer: for the word "addError" API, can we format that as the code font (like how 'session.id' is formatted for example)

Author: Yep, thanks for catching this

Errors can be viewed across both RUM and Error Tracking. Source and custom errors will be processed by Error Tracking, while console errors will be solely in RUM.

#### Sample search: List of all crashes that occurred on the "CheckoutViewController"page in the iOS Shop.ist application

> In this example, search within the errors event for the application ID, crashes, and view (page) name to see all crashes that occurred on the "CheckoutViewController" page for a particular operating system and application.

Reviewer: left this suggestion a few min ago on google doc but for 61, it's a big long and isn't super clear. how about this:
_In this example, search within the errors event to see all crashes that occurred on the “CheckoutViewController” page for a particular operating system  and application. _

#### Sample search: A list of all images loaded on the /cart view, where the image size was greater than or equal to 1000 kilobytes (or 1MB)

In this example, select **Resources** from the event type dropdown, then make a query for images that loaded on the cart view and were larger or equal to 1000 kilobytes.

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-resources.png" alt="Diagram of the RUM event hierarchy." style="width:80%;">}}

### Long tasks
Long tasks are anything that blocks the UI thread for a given period of time. On mobile, for example, a long task may be a frozen frame if the screen is blocked for longer than 300 ms.

#### Sample search: All frozen frame long tasks that lasted more than 500 ms

In this example, select Long tasks from the event type dropdown and specify the duration.

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-long-tasks.png" alt="Diagram of the RUM event hierarchy." style="width:80%;">}}

> ## No data appears after writing a query

Reviewer: i felt like going into this part of the guide as a bit abrupt -- can we add some extra spacing or a line of some sort (or even verbiage) to indicate we are now transitioning to a troubleshooting part of the guide?

Author: Good point. Added a troubleshooting header.

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-no-data-appears.png" alt="Diagram of the RUM event hierarchy." style="width:80%;">}}

If you aren't seeing data after writing a query, confirm that the event selector matches what you have in the search bar. In the example above, the event selector is set to search within **views**, but the search bar only contains  **action** attributes. To view action-related data, switch the view selector to actions. If you still don't see any data, check the time picker to ensure you are in a time window where data should be appearing.

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-no-data-appears-2.png" alt="Diagram of the RUM event hierarchy." style="width:80%;">}}

## How to query a event type that is nested in a different event type 

> When querying for specific actions, you can use the parent event type, but not one at equal or lower level. For example, actions are nested underneath views, and actions and errors are at the same level in the hierarchical chain. This means you can query for all actions and events that happened on a given page, but not for all actions that had a specific error type.

Reviewer: i think in the last part of the sentence, instead of the word "events" it should be errors, right?

Author: Yes, updated

> #### Sample search: The top 10 actions that occurred on / (the homepage)

Reviewer: can "/" be in code formatting? i also don't know if we need to write "the homepage"

> In this example, search within the actions event type for all view names in the Top List view to see the top 10 actions that occurred on /, which represents the homepage.

Reviewer: same comment as above

  tag: 'Documentation'
  text: 'Learn about the RUM Explorer'
- link: '/real_user_monitoring/'
  tag: 'Documentation'
  text: 'Learn how to visualize your RUM data'
---

## Overview

This guide walks through the different [types of data][1] that RUM collects and describes the hierarchy of each event type. See the diagram below for a high level overview of the hierarchy.

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-session-hierarchy-overview.png" alt="Diagram of the RUM event hierarchy." style="width:50%;">}}

## Sessions
All RUM data refers to user or synthetics sessions, which are at the top of the event hierarchy. A session is a unique user journey and encompasses everything (e.g., pages viewed, views, clicks, scrolls, errors) the user triggered. A session can last up to 4 hours of continuous activity, or it can expire after [15 minutes of inactivity][2]. Since a session encompasses the entire journey, all [attributes][3] tied to that user are also tied to that session. For example, you may want to query on a default attribute, like `application.id`, then add something more custom, like [user attributes][4].

> #### Sample search: List all sessions from user "Jane Doe"

Reviewer 2: ```suggestion
#### Sample search: List all sessions from a user
```

Reviewer: i would be ok with (in general) getting rid of the sample search line if it's redundant

To list all sessions from a specific user, select **Sessions** from the event type dropdown, then make a search query for the user and application.

> {{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-sample-search-list-all-sessions.png" alt="Sample search listing all sessions from user "Jane Doe"." style="width:80%;">}}

Reviewer 2: It would be great if we could make this example more modern and update the name being used here to something other than Jane Doe, which is used for unidentified dead bodies. Not a blocker but, ideally we'd use something else. (Nonbinary names are also nice!))

Notice that each session is automatically associated with a unique `session.id`.

## Views
Within a session, a view event is created each time a user navigates to a page (Browser RUM) or screen/screen segment (Mobile RUM) of an application. 

Each view automatically collects multiple view specific attributes and data, such as text in the URL and timing metrics, like the load time of a given page. When querying for specific views, you can add any default level attributes, like device, operating system, user information, etc. However, for event-specific attributes, these must be view-specific. To switch over to view events only, you can adjust the event selector as shown in the video below.

> {{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-switch-views.mp4" alt="RUM views" video="true" style="width:80%;">}}

Reviewer 2: This doesn't seem super necessary as a mp4 file, especially given we have images of this drop down later on in the page.

Similarly to the `session.id`, each view automatically has a unique `view.id` connected to it. 

### Actions, errors, resources, and long tasks

> Within views, the SDK creates more granular events that all fall along the same level hierarchically. However, each event is unique and carries its own attributes and properties.

Reviewer 2: ```suggestion
Within views, the SDK creates more granular events that all fall along the same hierarchy level. However, each event is unique and carries its own attributes and properties.
```

### Actions

Actions represent user activity on a page. In browsers, all click actions are automatically collected. On mobile, all taps, swipes, and scrolls are collected. Beyond these default actions, you can also send [custom actions][5], such as form completion and business transactions. 

> #### Sample search: Top list of all "Add to cart" actions that led to an error

Reviewer 2: This header feels kind of redundant given the text is almost identical on the next line
```suggestion
#### Sample search: Top list of actions leading to an error
```

> In this example, search for all actions from users clicking on the "Add to cart" button that resulted in an error.

Reviewer 2: ```suggestion
This example displays a query that searches for all actions from users clicking on the "Add to cart" button that resulted in an error.
```

> {{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-actions-all-add-to-cart.png" alt="Sample search of all "Add to Cart" actions that led to an error." style="width:80%;">}}

Reviewer 2: ```suggestion
{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-actions-all-add-to-cart.png" alt="Sample search of all 'Add to Cart' actions that led to an error." style="width:80%;">}}
```

### Errors

> You can use RUM to collect [frontend errors][6] that occur during the user session. By default, the browser SDK creates error events for unhandled exceptions and console errors. Additionally, you can collect custom errors through the RUM `addError` API ([browser][7] | [mobile][8]). On mobile apps, you can also see if the error led to a session termination, also known as a crash.

Reviewer 2: ```suggestion
You can use RUM to collect [frontend errors][6] that occur during the user session. By default, the browser SDK creates error events for unhandled exceptions and console errors. Additionally, you can collect custom errors through the RUM `addError` API (on [browser][7] and [mobile][8]). On mobile apps, you can also see if the error led to a session termination, also known as a crash.
```

> Errors can be viewed across both RUM and Error Tracking. Source and custom errors will be processed by Error Tracking, while console errors will be solely in RUM.

Reviewer 2: ```suggestion
Errors can be viewed in both RUM and Error Tracking. Source and custom errors are processed by Error Tracking, while console errors are solely in RUM.
```

> #### Sample search: List of all crashes that occurred on the "CheckoutViewController"page in the iOS Shop.ist application

Reviewer 2: ```suggestion
#### Sample search: List of all crashes that occurred on a page in the application
```

> In this example, search within the errors event to see all crashes that occurred on the “CheckoutViewController” page for a particular operating system and application.

Reviewer: we can cut out "operating system". it's not actually in the query from the screenshot

> In this example, search within the errors event to see all crashes that occurred on the “CheckoutViewController” page for a particular operating system and application.

Reviewer 2: ```suggestion
This example displays query that searches within the errors event to view all crashes that occurred on the “CheckoutViewController” page for a particular application.
```

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-sample-search-checkoutviewcontroller.png" alt="Sample search of all crashes that occurred on a page." style="width:80%;">}}

### Resources

> Resources are collected from views and include external requests from your application to a network provider, like XHR, JS loading, images, fonts, etc. Since it is collected from a view, you can query for all resources loaded on an application, or scope it down to just resources that occurred in a single view. 

Reviewer 2: ```suggestion
Resources are collected from views and include external requests from your application to a network provider, things like XHR, JS loading, images, or fonts, for example. Since it is collected from a view, you can query for all resources loaded on an application, or scope it down to just resources that occurred in a single view. 
```

> #### Sample search: A list of all images loaded on the /cart view, where the image size was greater than or equal to 1000 kilobytes (or 1MB)

Reviewer 2: ```suggestion
#### Sample search: A list of all images loaded on the `/cart` view filtered by image size
```

> In this example, select **Resources** from the event type dropdown, then make a query for images that loaded on the cart view and were larger or equal to 1000 kilobytes.

Reviewer 2: ```suggestion
In this example, **Resources** is selected from the event type dropdown, then a query for images that loaded on the cart view and were larger or equal to 1000 kilobytes are listed.
```

> {{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-resources.png" alt="Sample search of all images loaded on the cart view that were 1 megabyte or greater." style="width:80%;">}}

Reviewer 2: ```suggestion
{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-resources.png" alt="Sample search of all images loaded on the cart view that were 1000 kilobytes or greater." style="width:80%;">}}
```

### Long tasks

> Long tasks are anything that blocks the UI thread for a given period of time. On mobile, for example, a long task may be a frozen frame if the screen is blocked for longer than 300 ms.

Reviewer 2: ```suggestion
Long tasks are any task that blocks the UI thread for a specified period of time. On mobile for example, a long task may be a frozen frame if the screen is blocked for longer than 300 milliseconds.
```

#### Sample search: All frozen frame long tasks that lasted more than 500 ms

> In this example, select Long tasks from the event type dropdown and specify the duration.

Reviewer 2: ```suggestion
In this example, **Long tasks** is selected from the event type dropdown and then duration is specified.
```

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-long-tasks.png" alt="Sample search of all frozen frame long tasks lasting longer than 500 milliseconds." style="width:80%;">}}

## Troubleshooting

### No data appears after writing a query

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-no-data-appears.png" alt="Example of no data appearing after writing a query." style="width:80%;">}}

> If you aren't seeing data after writing a query, confirm that the event selector matches what you have in the search bar. In the example above, the event selector is set to search within **views**, but the search bar only contains  **action** attributes. To view action-related data, switch the view selector to actions. If you still don't see any data, check the time picker to ensure you are in a time window where data should be appearing.

Reviewer 2: I don't think we refer to it as a time picker but also, I don't see any kind of time frame selection in the screenshot above, which is confusing.

{{< img src="real_user_monitoring/guide/understanding-rum-event-hierarchy/rum-no-data-appears-2.png" alt="Example of updating a query by using the view selector." style="width:80%;">}}

> ### How to query a event type that is nested in a different event type 

Reviewer 2: ```suggestion
### Querying an event type that is nested in a different event type 
```

When querying for specific actions, you can use the parent event type, but not one at equal or lower level. For example, actions are nested underneath views, and actions and errors are at the same level in the hierarchical chain. This means you can query for all actions and errors that happened on a given page, but not for all actions that had a specific error type.

#### Sample search: The top 10 actions that occurred on `/`

> In this example, search within the actions event type for all view names in the Top List view to see the top 10 actions that occurred on `/`, which represents the homepage.

Reviewer 2: ```suggestion
This example searches within the actions event type for all view names using the Top List view to see the top 10 actions that occurred on `/`, which represents the homepage.
```
